bigflow/dataproc.py

```diff
@@ -91,23 +94,26 @@ def _ensure_has_setup_file(self):
             self._project,
             self._any_file_inside_project)
 
-    def _generate_internal_jobid(self, runtime):
+    def _generate_internal_jobid(self, context):
         job_random_suffix = ''.join(random.choices(string.ascii_lowercase + string.digits, k=10))
         job_timestamp = datetime.datetime.now().strftime("%Y-%m-%d-%H%M%S")  # FIXME: use 'runtime' ?
         jobid = f"{self.id.lower()}-{self.env or 'none'}-{job_timestamp}-{job_random_suffix}"
         logger.info("Internal job is %r", jobid)
         return jobid
 
-    def _prepare_driver_callable(self, runtime):
-        kwargs = dict(self.driver_arguments)
-        kwargs['runtime'] = runtime
-        
-        envs = {}
-        if self.env:
-            logger.debug("Will set env variable 'env' on pyspark driver to %r", self.env)
-            envs['env'] = self.env
+    def _prepare_driver_callable(self, context: bigflow.JobContext):
+        # Remove 'workflow' from context but keeps 'workflow_id'.
+        # In most cases Workflow is not pickleable as it keeps references to custom user jobs.
+        cleaned_context = context._replace(workflow=None)
```

Comment:
Maybe you can add something like `PickableContext` which takes a standard context in the init. It clarifies the intention and allows you to avoid mutating the context parameter.

Reply:
Idea was to pass this context (JobContext) to user-defined callback functiton.
Now it has signature: typing.Callback[JobContext]

I don't think there is real need to add something extra (some additional class) to such api, as 'workflow' is already Optional[]. As an option we can recreate 'workflow' object during driver unpickling, but IMO this is an overkill for now.

Reply:
ok 👍